[SPARK-27112][CORE] : Create a resource ordering between threads to resolve the deadlocks encountered …

pgandhi999 opened the pull request:
…when trying to kill executors either due to dynamic allocation or blacklisting

## What changes were proposed in this pull request?

There are two deadlocks as a result of the interplay between three different threads:

**task-result-getter thread**

**spark-dynamic-executor-allocation thread**

**dispatcher-event-loop thread(makeOffers())**

The fix ensures ordering synchronization constraint by acquiring lock on `TaskSchedulerImpl` before acquiring lock on `CoarseGrainedSchedulerBackend` in `makeOffers()` as well as killExecutors() method. This ensures resource ordering between the threads and thus, fixes the deadlocks.

## How was this patch tested?

Manual Tests

Commit:
[SPARK-27112] : Spark Scheduler encounters two independent Deadlocks when trying to kill executors either due to dynamic allocation or blacklisting

Ordered synchronization constraint by acquiring lock on Task Scheduler before acquiring lock on CoarseGrainedSchedulerBackend

AmplabJenkins commented:
Can one of the admins verify this patch?

pgandhi999 commented:
ok to test

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #103386 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103386/testReport)** for PR 24072 at commit [`e649900`](https://github.com/apache/spark/commit/e649900200bc18a57f0f60dafdda34b2acb30a1e).

attilapiros commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -631,8 +635,10 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       force: Boolean): Seq[String] = {
     logInfo(s"Requesting to kill executor(s) ${executorIds.mkString(", ")}")
 
+    val idleExecutorIds = executorIds.filter { id => force || !scheduler.isExecutorBusy(id) }
```
Nit: I would not use the name `idleExecutorIds` for this variable as when flag `force` is true then not only idle executors are contained.

squito replied:
meh, I'm not sure what else you'd call it ... there is already `executorsToKill` lower down ... unless you have a better suggestion, `idleExecutorIds` is probably good enough

but this does leave a small race for SPARK-19757, doesn't it?  After this executes, then an executor gets a task scheduled on it so its no longer idle, but you still kill it below?  To really prevent that, you'd need to get both locks (in the same order of course) so

```
val response = scheduler.synchronized { this.synchronized {
```

it also isn't the worst thing in the world if we occasionally kill an executor which just got a task scheduled on it.

abellina replied:
@squito  If you wanted to prevent that then you need to put the `val idleExecutorIds = executorIds.filter { id => force || !scheduler.isExecutorBusy(id) }` back inside that scheduler lock, right?

> it also isn't the worst thing in the world if we occasionally kill an executor which just got a task scheduled on it.

So we don't count this as a task failure right? Not sure where to look to verify that.

squito replied:
yes, I meant with the filter happening inside both locks -- more like it was before the current form of the PR, or as you suggested

attilapiros replied:
I have a suggestion for naming but I do not insist on that:
- renaming  `idleExecutorIds ` to `executorsToKill`
- renaming the old `executorsToKill` to `knownExecutorsToKill`

I also have checked the `synchronised` blocks of `CoarseGrainedSchedulerBackend` and its derived classes and have not found any other place where the scheduler is used for locking (within the synchronised block).

squito commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -258,15 +258,17 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
-              Some(executorData.executorAddress.hostPort))
-        }.toIndexedSeq
-        scheduler.resourceOffers(workOffers)
+      val taskDescs = scheduler.synchronized {
```
there should be a comment here about why we need both of these locks.

squito reviewed (commented):
I think this makes sense, but there are more instance of the order inversion in the locks.  I noticed at least `CoarseGrainedSchedulerBackend.disableExecutor()` also reverses the lock order.

abellina commented:
@squito the `CoarseGrainedSchedulerBackend.disableExecutor()`, I don't think it inverts the order. the `scheduler.executorLost` call happens outside of the `CoarseGrainedSchedulerBackend` lock, and it gets called from a driver event loop. Let me know if I've missed something.

vanzin commented:
Your PR title and description are basically copies of the bug. Could you instead describe the change?

SparkQA commented:
**[Test build #103386 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103386/testReport)** for PR 24072 at commit [`e649900`](https://github.com/apache/spark/commit/e649900200bc18a57f0f60dafdda34b2acb30a1e).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103386/
Test FAILed.

squito commented:
@abellina you're right about `disableExecutors`, thanks for taking a closer look, sorry that was just from a really quick scan.  But we should be sure to take a close look at all places the lock is used.

pgandhi999 commented:
> but this does leave a small race for SPARK-19757, doesn't it? After this executes, then an executor gets a task scheduled on it so its no longer idle, but you still kill it below? To really prevent that, you'd need to get both locks (in the same order of course) so
> 
> val response = scheduler.synchronized { this.synchronized {

@squito I did think about this yesterday and tried it out as well; the deadlock issue gets fixed alongwith the race, but I was not sure whether doing this may or may not cause a performance degradation as a bunch of threads might end up busy waiting a lot of time. I can do some perf tests with the above change and if it looks good, update the PR with the fix. Will let you know. Thank you.

dhruve commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -641,7 +647,6 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
       // If this executor is busy, do not kill it unless we are told to force kill it (SPARK-9552)
```
Remove this comment and add one where we are doing the `force` check.

Commit:
[SPARK-27112] : Addressing Reviews March 13, 2019

Locking the code block in killExecutors() method with TaskSchedulerImpl followed by CoarseGrainedSchedulerBackend to avoid race condition issue and adding comments.

SparkQA commented:
**[Test build #103453 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103453/testReport)** for PR 24072 at commit [`ed12daf`](https://github.com/apache/spark/commit/ed12daf28a09e425d71352b45edbdc15ad2b4cec).

vanzin commented:
Sorry to be a pain about this, but please remove the bug stuff from the PR description. If we want details about the bug, we can look at, ahem, the bug. Focus on describing what the change does and why it fixes the problem.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -258,15 +258,20 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+      // SPARK-27112: We need to ensure that there is ordering of lock acquisition
+      // between TaskSchedulerImpl and CoarseGrainedSchedulerBackend objects in order to fix
+      // the deadlock issue exposed in SPARK-27112
+      val taskDescs = scheduler.synchronized {
```
I took a quick look at the code that calls this, and I'm wondering if holding the two locks here is really needed.

For context, all this code is inside the RPC endpoint handler. This is a `ThreadSafeRpcEndpoint` so there's only one message being processed at a time, meaning that you won't have multiple threads calling `makeOffers` concurrently.

So it seems to me that it would be possible to:

- with the `CoarseGrainedSchedulerBackend.this` lock held, calculate the works offers.

```
val workOffers = CoarseGrainedSchedulerBackend.this.synchronized {
  ...
}
```

With the scheduler lock held, make the offers:

```
val taskDesc = scheduler.synchronized {
  scheduler.resourceOffers(workOffers)
}
```

And as far as I understand that should work and also be easier to understand, right?

I also noticed that later this code calls `launchTasks`, and that method accesses and modifies data in `executorDataMap` without the `CoarseGrainedSchedulerBackend.this` lock, which is very sketchy.

vanzin replied:
Ok, seems both locks are needed because of SPARK-19757. But the `launchTasks` issue is still there.

pgandhi999 replied:
@vanzin  So in the code, I came across the following comment, wonder if that answers the `launchTasks` issue. I exactly do not understand the intention of the comment though.

```
// Accessing `executorDataMap` in `DriverEndpoint.receive/receiveAndReply` doesn't need any
  // protection. But accessing `executorDataMap` out of `DriverEndpoint.receive/receiveAndReply`
  // must be protected by `CoarseGrainedSchedulerBackend.this`. Besides, `executorDataMap` should
  // only be modified in `DriverEndpoint.receive/receiveAndReply` with protection by
  // `CoarseGrainedSchedulerBackend.this`.
  private val executorDataMap = new HashMap[String, ExecutorData]
```

vanzin replied:
Ok, I think that's fine. I checked and all modifications happen on the endpoint thread, so reading the map from that thread without a lock should be fine. The data being modified (`freeCores`) is also only used in the endpoint thread, so that looks safe too.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -631,29 +641,33 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+        // If this executor is busy, do not kill it unless we are told to force kill it (SPARK-9552)
+        val executorsToKill = knownExecutors
+          .filter { id => !executorsPendingToRemove.contains(id) }
+          .filter { id => force || !scheduler.isExecutorBusy(id) }
```
In a similar vein to my previous comment, although I'm less sure about this one.

This seems to be the only interaction with the scheduler in this method, so could this filtering be done first thing in the method, with the scheduler lock held, and then the rest of the code just needs the `CoarseGrainedSchedulerBackend` lock?

It seems to me the behavior wouldn't change from the current state (where the internal scheduler state can change while this  method is running). And as in the other case, easier to understand things when you're only holding one lock.

vanzin replied:
(Caught up with the previous discussion and it seems that here both locks are needed to avoid an edge case where you could kill active executors.)

vanzin commented:
BTW if the "two locks need to be held" thing is really needed in multiple places, might be good to have a helper function, e.g.

```
def withLock[T](fn: => T): T = lock1.synchronized { lock2.synchronized { fn } }
```

Commit:
[SPARK-27112] : Adding a helper function

SparkQA commented:
**[Test build #103514 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103514/testReport)** for PR 24072 at commit [`2b4f226`](https://github.com/apache/spark/commit/2b4f226f365dc769562e0c0e99b048201f1e398f).

pgandhi999 commented:
@squito @vanzin @attilapiros @abellina Have worked on all the comments and pushed the respective changes.

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -258,7 +258,10 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
     private def makeOffers() {
       // Make sure no executor is killed while some task is launching on it
-      val taskDescs = CoarseGrainedSchedulerBackend.this.synchronized {
+      // SPARK-27112: We need to ensure that there is ordering of lock acquisition
```
This comment would be great in the `withLock` function, instead of being copy & pasted in a few places.

pgandhi999 replied:
Done

vanzin commented on core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala:
```diff
@@ -258,7 +258,10 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
+      // SPARK-27112: We need to ensure that there is ordering of lock acquisition
+      // between TaskSchedulerImpl and CoarseGrainedSchedulerBackend objects in order to fix
+      // the deadlock issue exposed in SPARK-27112
+      val taskDescs = withLock({
```
No need for the parentheses.

pgandhi999 replied:
Done

pgandhi999 replied:
Sorry for the extra commits, was fixing code indentation.

Commit:
[SPARK-27112] : Writing comment in single place and removing parenthesis

SparkQA commented:
**[Test build #103518 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103518/testReport)** for PR 24072 at commit [`47448b7`](https://github.com/apache/spark/commit/47448b794b9521a638ba44b3396e27a42c580362).

Commit:
[SPARK-27112] : Formatting Code

Commit:
[SPARK-27112] : Indenting code

SparkQA commented:
**[Test build #103519 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103519/testReport)** for PR 24072 at commit [`09f9b47`](https://github.com/apache/spark/commit/09f9b4767b3f8b94b8ef1ae956d46e7158d50b9d).

SparkQA commented:
**[Test build #103518 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103518/testReport)** for PR 24072 at commit [`47448b7`](https://github.com/apache/spark/commit/47448b794b9521a638ba44b3396e27a42c580362).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103518/
Test FAILed.

SparkQA commented:
**[Test build #103514 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103514/testReport)** for PR 24072 at commit [`2b4f226`](https://github.com/apache/spark/commit/2b4f226f365dc769562e0c0e99b048201f1e398f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103514/
Test PASSed.

SparkQA commented:
**[Test build #103519 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103519/testReport)** for PR 24072 at commit [`09f9b47`](https://github.com/apache/spark/commit/09f9b4767b3f8b94b8ef1ae956d46e7158d50b9d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103519/
Test PASSed.

attilapiros commented:
LGTM

squito commented:
lgtm

abellina reviewed (approved):
👍

dhruve commented:
+1
@squito @vanzin Can we merge this PR in. Thanks.

squito commented:
merged to master.

@pgandhi999 there was a merge conflict against branch-2.4, would you mind opening another PR against that branch?

pgandhi999 commented:
Sure @squito will do that. Thank you.
